portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java

```diff
@@ -0,0 +1,592 @@
+				Arrays.asList(portlet), URLType.JAVASCRIPT);
+		}
+
+		return new PortletRenderParts(
+			footerCssPaths, footerJavaScriptPaths, headerCssPaths,
+			headerJavaScriptPaths, portletHTML, !portlet.isAjaxable());
+	}
+
+	private static String _getRootPortletId(Portlet portlet) {
+
+		// Workaround for portlet#getRootPortletId because that does not return
+		// the proper root portlet ID for OpenSocial and WSRP portlets
+
+		Portlet rootPortlet = portlet.getRootPortlet();
+
+		return rootPortlet.getPortletId();
+	}
+
+	private static Collection<String> _getStaticURLs(
+		HttpServletRequest httpServletRequest,
+		Collection<PortletResourceAccessor> portletResourceAccessors,
+		Collection<Portlet> portlets, Set<String> visitedURLs) {
+
+		List<String> urls = new ArrayList<>();
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		for (Portlet portlet : portlets) {
+			Portlet rootPortlet = portlet.getRootPortlet();
+
+			for (PortletResourceAccessor portletResourceAccessor :
+					portletResourceAccessors) {
+
+				String contextPath = null;
+
+				if (portletResourceAccessor.isPortalResource()) {
+					contextPath = PortalUtil.getPathContext();
+				}
+				else {
+					contextPath =
+						PortalUtil.getPathProxy() + portlet.getContextPath();
+				}
+
+				Collection<String> portletResources =
+					portletResourceAccessor.get(portlet);
+
+				for (String portletResource : portletResources) {
+					boolean module = false;
+
+					if (portletResource.startsWith("module:")) {
+						module = true;
+
+						portletResource = portletResource.substring(7);
+					}
+
+					if (!HttpComponentsUtil.hasProtocol(portletResource)) {
+						portletResource = PortalUtil.getStaticResourceURL(
+							httpServletRequest, contextPath + portletResource,
+							rootPortlet.getTimestamp());
+					}
+
+					if (!portletResource.contains(Http.PROTOCOL_DELIMITER)) {
+						String cdnBaseURL = themeDisplay.getCDNBaseURL();
+
+						portletResource = cdnBaseURL.concat(portletResource);
+					}
+
+					if (module) {
+						portletResource = "module:" + portletResource;
+					}
+
+					if (visitedURLs.contains(portletResource)) {
+						continue;
+					}
+
+					visitedURLs.add(portletResource);
+
+					urls.add(portletResource);
+				}
+			}
+		}
+
+		return urls;
+	}
+
+	private static Collection<String> _getURLs(
+		HttpServletRequest httpServletRequest,
+		Collection<PortletResourceAccessor> portletResourceAccessors,
+		Collection<Portlet> portlets, URLType urlType) {
+
+		boolean fastLoad;
+		Predicate<String> predicate = null;
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		if (urlType == URLType.CSS) {
+			fastLoad = themeDisplay.isThemeCssFastLoad();
+		}
+		else if (urlType == URLType.JAVASCRIPT) {
+			fastLoad = themeDisplay.isThemeJsFastLoad();
+
+			predicate = resource -> !themeDisplay.isIncludedJs(resource);
```

Comment:
I've seen this `predicate` before, but I really have no idea what is does. Could you explain it very briefly or link me to something about the purpose of `predicate`?